### ISSUE TITLE
Static site generator?

### ISSUE THREAD

**Issue by zenflow** [sweetalert2/sweetalert2.github.io#2]
Anyone thought about using a [static site generator](https://www.staticgen.com/) for this? 

It might help with a few things:
- Some (not all) content in Markdown (easier to write and review than html)
- Automatic generation of navigation component from definition of sections/pages
- Automatic generation of sitemap.xml from definition of sections/pages
- Automatically add "Edit this document" on each section/page, linking to that particular file on Github
- ...

It is kind of overwhelming how many options there are though...	We can probably limit our options to those projects that are for Node.js.

Is this a good idea? If so, can anyone recommend a particular static site generator?

**Comment by acupajoe** [comment]
Definitely right that there are a ton of options out there! I haven't personally needed to do static site generation before, but I'd be willing to learn [Gatsby](https://github.com/gatsbyjs/gatsby).

Did you have other ideas? (Sorry on the delay getting to this)

**Comment by zenflow** [comment]
> Did you have other ideas?

Not really. Gatsby caught my eye as well since it is a Node.js project, uses React for it's templating, and is the most popular of it's kind (meaning a relatively rich ecosystem & community).

Interestingly, [Nuxt.js](https://github.com/nuxt/nuxt.js) is listed on staticgen.com, but [Next.js](https://github.com/zeit/next.js) is not. I am getting into Next.js; it is used for dynamic Server-Side-Rendered sites, but it can also be used for generating static sites. It's probably not as good an option as Gatsby though since it (and it's community & ecosystem) are not specialized on static sites.

**Comment by zenflow** [comment]
No worries on the delay. We all just do what we can when we can.

**Comment by acupajoe** [comment]
Seems like Gatsby is the best bet for what we're looking for! I'll be messing around with it this week to figure out how it works.

**Comment by zenflow** [comment]
@acupajoe Did you have a chance to fool around with Gatsby yet? How does it look?

**Comment by acupajoe** [comment]
@zenflow I am currently! Getting the current page reimplemented in Gatsby on a local branch at the moment. I plan on having it done either this afternoon or tomorrow afternoon

**Comment by acupajoe** [comment]
@zenflow I don't have access to this repo to be able to push my branch to it, could you fix that please?

**Comment by limonte** [comment]
@acupajoe you're promoted to the owner of the SweetAlert2 organization :sunglasses:

**Comment by acupajoe** [comment]
@zenflow if you'll check the `dev/gatsby` branch - you'll see at least the beginnings of what the static site generation looks like. 

Things I'm finding frustrating at the moment:
- `<pre>` tags are annoying to work with inside of a react style application (spans aren't working correctly and indentation makes the code look weird)
- We'll have to move all the frontend event code into handlers into jsx which will take some time.

Care to take a look?

**Comment by zenflow** [comment]
Also would be awesome to use markdown as the source for the bulk of the documentation pages, and apply all the above mentioned features features to:
```js
// js blocks
```

**Comment by acupajoe** [comment]
Ah yes, time to rethink indeed! Wanted to get my feet wet with Gatsby before jumping too far in. Incremental goals and whatnot.

Here's what we've got for the redesign:
- Separation of content by sections all on a single page
- Implementation of codemirror type examples (codepen.io would also work and I've reached out to them to see if they offer packages for open-source teams, so that is an option!)
- Handle documentation in a markdown friendly way
- Provide syntax highlighting
- New design, which currently is the image I have from [#711](https://github.com/sweetalert2/sweetalert2/issues/711)

**Comment by zenflow** [comment]
> codepen.io would also work and I've reached out to them to see if they offer packages for open-source teams, so that is an option

Do you mean [embedding codepen examples](https://codepen.io/embeds) in our site? (Otherwise could you link me to the product page?) I think it would be preferable if the code samples live in our repo, and it seems like codepens live on codepen.io.

**Comment by acupajoe** [comment]
Yeah that was going to be the idea, but I agree, hosting our own code would be preferable.

**Comment by zenflow** [comment]
This might be a good place to start: https://github.com/ericwindmill/gatsby-starter-docs

Looks like it has a bunch of our desired features already (structure for TOC and sitemap, content in markdown, syntax highlighting with Prism.js), and more. 

Changes we would need to make
- All sections on a single page (to keep ctrl+f ability)
- Button to execute code samples (and maybe some other fancy buttons, like "Open in codepen/jsfiddle")
- The visual design (obviously), but I think it should not be difficult to start with this (?)

**Comment by zenflow** [comment]
> Separation of content by sections all on a single page
> All sections on a single page (to keep ctrl+f ability)

Instead of this, we could use [docsearch.js](https://www.npmjs.com/package/docsearch.js)

**Comment by acupajoe** [comment]
I love docsearch and it seems to be very well adopted throughout the community as well (vue, react, babel, etc.)! Good idea

**Comment by limonte** [comment]
> I love docsearch and it seems to be very well adopted throughout the community as well (vue, react, babel, etc.)!

I just checked vue, react, babel docs pages and all of them are using [Algolia](https://www.algolia.com/):

![](https://i.imgur.com/heFf5BH.png)

![](https://i.imgur.com/w1PoNU3.png)

![](https://i.imgur.com/pGGvDsL.png)

But IMO both docsearch.js and Algolia are overkill for SweetAlert2. Old good ctrl+f will work just fine for the amount of information we are planning to have.

**Comment by acupajoe** [comment]
Ah yeah @limonte docsearch is by algolia - will we not want to add examples or anything at some point?

**Comment by zenflow** [comment]
> But IMO both docsearch.js and Algolia are overkill for SweetAlert2. Old good ctrl+f will work just fine for the amount of information we are planning to have

@limonte The benefit of docsearch.js would be that we can have the sections on separate pages (seems like the common setup, but doesn't allow you to ctrl+f) and still have the ability to search. So the amount of content doesn't really factor into it, unless you consider that there is enough content that we want to divide it up.

**Comment by zenflow** [comment]
> This might be a good place to start

On second thought, it might be better if we start fresh and just use the ideas (organization, plugins, code) from this repo and others.

**Comment by zenflow** [comment]
> Button to execute code samples

Runkit can do this (along with the syntax highlighting). [Examples can live in this repo](https://runkit.com/docs/embed) and [there is a React component we can use in Gatsby](https://github.com/runkitdev/react-runkit). But this does not work with the standard `gatsby-transformer-remark` package,.. yet. I think I will make a Gatsby plugin for this, unless anyone is opposed to this approach.

**Comment by zenflow** [comment]
So I really want to take care of this issue and #1 asap so that writing code examples is not such a pain.

I did some studying and playing around with Gatsby too, actually almost 2 months ago. I think it could be amazing if we had it set up but it's a bit complex and requires a lot of customization with different plugins, transformers, etc.

Recently I've been looking at [Docusaurus](https://docusaurus.io), which is also in the top 10 on www.staticgen.com, also a Node.js project, and also uses React for rendering html. The important difference is that Docusaurus is not just a static site generator but a **documentation** site generator, so it will require much less customization. 

Anyone object to using Docusaurus for a new overhauled documentation site?

I'm checking now that it can fulfill all our requirements, and provided it does, I'm going to start making it right away. If it's not exactly what y'all would ultimately like, it will at least be a large improvement in what we're doing now, and hopefully we can use it until it's either replaced or improved. For now I won't worry about about most of the ideas we talked about in https://github.com/sweetalert2/sweetalert2/issues/711 .. we can improve the organization and content and visual design in separate PRs.. I'm just going to copy those things from the current site for now.

**Comment by limonte** [comment]
No objections for Docusaurus from my side, it looks nice and simple to use and maintain.

**Comment by zenflow** [comment]
https://github.com/sweetalert2/sweetalert2.github.io/pull/20#issuecomment-398709548 by @limonte 

> I changed my mind. Let's not require from our contributors the knowledge of React, Docusaurus, markdown, whatnot just because we can.
> 
> There are no real issues we will solve with bringing new "cool" technologies. Nothing is better than good old HTML.
> 
> The current solution is well tested and works perfectly on all devices, I don't have time to do the same work for the new solution. Contributors seem to be busy, I can't and I shouldn't count on them.
> 
> [React isn't easy](http://bradfrost.com/blog/post/my-struggle-to-learn-react/) even if some of us feel so. The current state of the front-end is terrible. Let's admit it and patiently wait for the bright future. Until then, 
> 
> <img src="https://user-images.githubusercontent.com/6059356/41654132-1bbc1564-7491-11e8-9089-e5029874edfd.png" width="300">

**Comment by zenflow** [comment]
@limonte 

> Let's not require from our contributors the knowledge of React, Docusaurus, markdown [...]

Contributors wouldn't need any knowledge of React or Docusaurus (unless they want to change the landing page).. just markdown.. is knowledge of markdown too much to require? I think you may have been too hasty in making this decision, and not taken the time to take in and consider the facts. 

> [...] just because we can
> There are no real issues we will solve with bringing new "cool" technologies.

Actually there's a ton of reasons to use a static site generator besides just "we can".... See the original comment. Also I've created some new issues that are key: #27 #28 #29. 

> Nothing is better than good old HTML.
> React isn't easy even if some of us feel so. The current state of the front-end is terrible. Let's admit it and patiently wait for the bright future.

React *is* the bright future. It's advantage over some other libraries that help make a **dynamic UI** is it's simplicity and the fact that it's largely just HTML in JS. If you know HTML and JS, you are not far from knowing React. Which brings me to my comment on the linked blog post: The author (by his own admission) is not a web *developer* and doesn't know JS very well (see his 6th point). This makes all the other points make more sense.

**Comment by limonte** [comment]
> React is the bright future.

I wouldn't be so sure about it. If so, why Vue exist and becoming [even more popular than React](https://twitter.com/dan_abramov/status/1007439168400654336), but still different in fundamental techniques? What if one of SweetAlert2's collaborators would be the experienced Vue developer and would insist on VuePress instead of Docusaurus? 

I personally think that both of them will be replaced by something better in the very near future. It's not clear for me with what, but both frameworks aren't ideal even if at the moment they are the best we have.

Sorry that I have more questions than answers, but that's how it is :) Let's stay with Vanilla JS for now and be carefree about the future.

**Comment by zenflow** [comment]
> I wouldn't be so sure about it. If so, why Vue exist and becoming even more popular than React, but still different in fundamental techniques?

Because TIMTOWTDI (there is more than one way to do it). Both are excellent and huge improvements over HTML and vanilla JS. You probably know this, but React is still more popular in terms of downloads, and probably in terms of actual usage too. I think it will continue to be pretty dominant and not be replaced in the near future.

> What if one of SweetAlert2's collaborators would be the experienced Vue developer and would insist on VuePress instead of Docusaurus?

I would be thrilled that (a) we can resolve all the issues with a manually written vanilla JS and HTML page, and (b) I can get some experience with Vue.

**Comment by zenflow** [comment]
This debate about React vs Vue vs nothing has detracted from my main point:

> Contributors wouldn't need any knowledge of React or Docusaurus (unless they want to change the landing page).. just markdown.. is knowledge of markdown too much to require? 

Is it?

**Comment by limonte** [comment]
> This debate about React vs Vue

It's not. It's about the fragility of both. Both are opinionated. Opinionated solutions are temporary and replacing them with new technologies is the painful process. Do you have experience with rewriting any project from cool-framework-a to cool-framework-b? I do and as you can see I'm traumatized by that :)

Requiring the markdown knowledge isn't a big deal of course.

**Comment by zenflow** [comment]
> It's not.

Ok.. the debate about React (or Vue) vs nothing (i.e. vanilla JS + HTML) you know what debate I mean. 

My main point was that your leading point for rejecting a static site generator is not valid: contributors only need knowledge of markdown. 

> Both are opinionated. Opinionated solutions are temporary and replacing them with new technologies is the painful process. Do you have experience with rewriting any project from cool-framework-a to cool-framework-b? I do and as you can see I'm traumatized by that

But this is just a documentation website, and a rewrite doesn't even need to be exactly the same; when something better comes along in a few years it wouldn't be a complicated task to copy and paste the content and make a new landing page. I.e. just build a new website, like I did in #20. I mean, you wouldn't even need to look at the source to do that, but the future solution should probably have the content in markdown also, which would make it even easier. Meanwhile we could take advantage of a better format for documentation.

**Comment by limonte** [comment]
Related React vs Platform discussion: https://github.com/mdn/sprints/issues/967 

It would be fun to see them in pain of rewriting everything in 3 years from React to `<insert_the_name_of_the_next_coolest_technology>`.
